### PR TITLE
KAFKA-14717 KafkaStreams can' get running if the rebalance happens be…

### PR DESCRIPTION
**Pull request by chia7712**
I noticed this issue when tracing #12590
StreamThread closes the consumer before changing state to DEAD. If the partition rebalance happens quickly, the other StreamThreads can't change KafkaStream state from REBALANCING to RUNNING since there is a PENDING_SHUTDOWN StreamThread

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/src/test/java/org/apache/kafka/streams/integration/AdjustStreamThreadCountTest.java [hunk 1]
```diff
@@ -257,6 +258,23 @@ private Thread adjustCountHelperThread(final KafkaStreams kafkaStreams, final in
         });
     }
 
+    @Test
+    public void testRebalanceHappensBeforeStreamThreadGetDown() throws Exception {
+        final Properties prop = new Properties();
+        prop.putAll(properties);
+        // make rebalance happen quickly
+        prop.put(ConsumerConfig.HEARTBEAT_INTERVAL_MS_CONFIG, 200);
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/integration/AdjustStreamThreadCountTest.java]: Is there a better way to test this scenario than relying on the real time (and hence is time dependendent)? More specifically, I'm looking for a test case which would 100% fail without the fix, and would 100% pass with the fix. While this test seems would be pass some times even without the fix, is that right?

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java [hunk 2]
```diff
@@ -658,6 +658,8 @@ public synchronized void onChange(final Thread thread,
                         setState(State.REBALANCING);
                     } else if (newState == StreamThread.State.RUNNING) {
                         maybeSetRunning();
+                    } else if (state != State.RUNNING && newState == StreamThread.State.DEAD) {
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: I wonder if it's cleaner to change the logic of `maybeSetRunning` directly, e.g. to exclude threads which are `PENDING_SHUTDOWN` as long as there are still other threads running?

**Review by guozhangwang — COMMENTED** [review]
Thanks @chia7712 for the find! After digging it I agree this is a real issue. Though I have some questions about the fix itself, and also about the test covering this scenario. Please let me know what do you think?

File: streams/src/test/java/org/apache/kafka/streams/integration/AdjustStreamThreadCountTest.java (hunk 1 above)
Comment (chia7712) [reply streams/src/test/java/org/apache/kafka/streams/integration/AdjustStreamThreadCountTest.java]: > While this test seems would be pass some times even without the fix, is that right?

you are right. I have updated the test to make sure it is always failed without the fix.

**Comment by guozhangwang** [comment]
LGTM. Merged to trunk.

**Pull request merged**